staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -238,43 +303,245 @@ func newAuthenticator(opts Options, initVerifier func(ctx context.Context, a *Au
+		client := &http.Client{Transport: r.t, Timeout: 30 * time.Second}
+		ctx := oidc.ClientContext(context.Background(), client)
+		av = newAsyncIDTokenVerifier(ctx, r.config, iss)
+		r.verifierPerIssuer[iss] = av
+	}
+	r.m.Unlock()
+
+	v := av.verifier()
+	if v == nil {
+		return nil, fmt.Errorf("verifier not initialized for issuer: %q", iss)
+	}
+	glog.V(4).Infof("Got verifier for: iss: %q: %+v", iss, v)
+	return v, nil
+}
+
+// expand extracts the distributed claims from claim names and claim sources.
+// The extracted claim value is pulled up into the supplied claims.
+//
+// Distributed claims are of the form as seen below, and are defined in the
+// OIDC Connect Core 1.0, section 5.6.2.
+// See: https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
+//
+// {
+//   ... (other normal claims)...
+//   "_claim_names": {
+//     "groups": "src1"
+//   },
+//   "_claim_sources": {
+//     "src1": {
+//       "endpoint": "https://www.example.com",
+//       "access_token": "f005ba11"
+//     },
+//   },
+// }
+func (r *claimResolver) expand(c claims) error {
+	const (
+		// The claim containing a map of endpoint references per claim.
+		// OIDC Connect Core 1.0, section 5.6.2.
+		claimNamesKey = "_claim_names"
+		// The claim containing endpoint specifications.
+		// OIDC Connect Core 1.0, section 5.6.2.
+		claimSourcesKey = "_claim_sources"
+	)
+
+	if glog.V(5) {
+		glog.Infof("initial claim set: %v", c)
+	}
+	defer func() {
+		if glog.V(4) {
+			glog.Infof("final claim set: %v", c)
+		}
+	}()
+	_, ok := c[r.claim]
+	if ok {
+		// There already is a normal claim, skip resolving.
+		return nil
+	}
+	names, ok := c[claimNamesKey]
+	if !ok {
+		// No _claim_names, no keys to look up.
+		return nil
+	}
+
+	claimToSource := map[string]string{}
+	if err := json.Unmarshal([]byte(names), &claimToSource); err != nil {
+		return fmt.Errorf("oidc: error parsing distributed claim names: %v", err)
+	}
+
+	rawSources, ok := c[claimSourcesKey]
+	if !ok {
+		// Having _claim_names claim,  but no _claim_sources is not an expected
+		// state.
+		return fmt.Errorf("oidc: no claim sources")
+	}
+
+	var sources map[string]endpoint
+	if err := json.Unmarshal([]byte(rawSources), &sources); err != nil {
+		// The claims sources claim is malformed, this is not an expected state.
+		return fmt.Errorf("oidc: could not parse claim sources: %v", err)
+	}
+
+	src, ok := claimToSource[r.claim]
+	if !ok {
+		// No distributed claim present.
+		return nil
+	}
+	ep, ok := sources[src]
+	if !ok {
+		return fmt.Errorf("oidc: malformed claim sources")
+	}
+	if ep.URL == "" {
+		// This is maybe an aggregated claim (ep.JWT != "").
+		return nil
+	}
+	return r.resolve(ep, c)
+}
+
+// resolve requests distributed claims from all endpoints passed in,
+// and inserts the lookup results into allClaims.
+func (r *claimResolver) resolve(endpoint endpoint, allClaims claims) error {
+	// TODO: cache resolved claims.
```

Comment:
Agreed, but it seemed like it would better be placed in a separate commit.  What do you advise? I'm fine with adding more code to this commit, but it's already large as it is.

Reply:
Add a private field to the config to let the tests enable this feature, and disable it by default?

Reply:
Good idea! Done.